File: ast/declaration/variable.go
package declaration

import (
	"github.com/panda-io/micro-panda/ast/core"
)

type Variable struct {
	DeclarationBase
	Typ   core.Type
	Const bool
	Value core.Expression

	Parent *Struct
}

func (v *Variable) IsConstant() bool {
	return v.Const
}

func (v *Variable) Kind() core.DeclarationKind {
	return core.DeclarationVariable
}

func (v *Variable) Type() core.Type {
	return v.Typ
}

func (v *Variable) ResolveType(c core.Context) {
	v.Typ = c.ResolveType(v.Typ)
}

func (v *Variable) Validate(c core.Context) {
	if v.Value != nil {
		v.Value.Validate(c, v.Typ)
	}
	if v.Const {
		if v.Value == nil {
			c.Error(v.GetPosition(), "const must be initialized when declare")
		} else if !v.Value.IsConstant() {
			c.Error(v.Value.GetPosition(), "expect const expression")
		}
	}
	if v.Value != nil {
		if v.Value.Type() == nil {
			c.Error(v.Value.GetPosition(), "unknown type")
		} else if !v.Value.Type().Equal(v.Typ) {
			c.Error(v.Value.GetPosition(), "init value type mismatch with define")
		}
	}
}

File: app/src/main/java/com/github/panda_io/micro_panda/ast/declaration/Variable.java
package com.github.panda_io.micro_panda.ast.declaration;

import com.github.panda_io.micro_panda.ast.type.Type;
import com.github.panda_io.micro_panda.ast.Context;
import com.github.panda_io.micro_panda.ast.expression.Expression;

public class Variable extends Declaration {
    public Type type;
    public boolean constant;
    public Expression value;
    public Struct parent;

    public boolean isConstant() {
        return this.constant;
    }

    public Type getType() {
        return this.type;
    }

    public void resolveType(Context context) {
        this.type = context.resolveType(this.type);
    }

    public void validate(Context context) {
        if (this.value != null) {
            this.value.validate(context, this.type);
        }
        if (this.constant) {
            if (this.value == null) {
                context.addError(this.getOffset(), "const must be initialized when declare");
            } else if (!this.value.isConstant()) {
                context.addError(this.getOffset(), "expect constant expression");
            }
        }
        if (this.value != null) {
            if (this.value.getType() == null) {
                context.addError(this.value.getOffset(), "unknown type");
            } else if (!this.value.getType().equal(this.type)) {
                context.addError(this.value.getOffset(), String.format("init value type mismatch with define, expect '%s', got '%s'", this.type.string(),
                this.value.getType().string()));
            }
            if (!this.value.isConstant()) {
                context.addError(this.value.getOffset(), "expect const expression");
            }
        }
    }
}
